File: src/main/java/org/nest/commons/_visitor/PowVisitor.java
package org.nest.commons._visitor;

import de.monticore.literals.literals._ast.ASTIntLiteral;
import org.nest.commons._ast.ASTExpr;
import org.nest.spl.symboltable.typechecking.Either;
import org.nest.symboltable.symbols.TypeSymbol;
import org.nest.units.unitrepresentation.UnitRepresentation;
import org.nest.utils.AstUtils;

import static de.se_rwth.commons.logging.Log.error;
import static org.nest.spl.symboltable.typechecking.TypeChecker.*;
import static org.nest.symboltable.predefined.PredefinedTypes.*;

/**
 * @author ptraeder
 */
public class PowVisitor implements CommonsVisitor{
  final String ERROR_CODE = "SPL_POW_VISITOR";

  @Override
  public void visit(ASTExpr expr){
    final Either<TypeSymbol, String> baseTypeE = expr.getBase().get().getType();
    final Either<TypeSymbol, String> exponentTypeE = expr.getExponent().get().getType();

    if (baseTypeE.isError()) {
      expr.setType(baseTypeE);
      return;
    }
    if (exponentTypeE.isError()) {
      expr.setType(exponentTypeE);
      return;
    }

    TypeSymbol baseType = baseTypeE.getValue();
    TypeSymbol exponentType = exponentTypeE.getValue();

    if (isNumeric(baseType) && isNumeric(exponentType)) {
      if (isInteger(baseType) && isInteger(exponentType)) {
        expr.setType(Either.value(getIntegerType()));
        return;
      }
      else if (isUnit(baseType)) {
        if (!isInteger(exponentType)) {
          final String errorMsg = ERROR_CODE+ " " + AstUtils.print(expr.get_SourcePositionStart()) + " : " +
              "With a Unit base, the exponent must be an integer.";
          expr.setType(Either.error(errorMsg));
          error(errorMsg,expr.get_SourcePositionStart());
          return;
        }
        UnitRepresentation baseRep = UnitRepresentation.getBuilder().serialization(baseType.getName()).build();
        Either<Integer, String> numericValue = calculateNumericValue(expr.getExponent().get());//calculate exponent value if exponent composed of literals
        if (numericValue.isValue()) {
          expr.setType(Either.value(getTypeIfExists((baseRep.pow(numericValue.getValue())).serialize()).get()));
          return;
        }
        else {
          final String errorMsg = numericValue.getError();
          expr.setType(Either.error(errorMsg));
          error(errorMsg,expr.get_SourcePositionStart());
          return;
        }
      }
      else {
        expr.setType(Either.value(getRealType()));
        return;
      }
    }
    //Catch-all if no case has matched
    final String errorMsg = ERROR_CODE+ " " + AstUtils.print(expr.get_SourcePositionStart()) + " : " +
        "Cannot determine the type of the expression: " + baseType.prettyPrint()
        +"**"+exponentType.prettyPrint();
    expr.setType(Either.error(errorMsg));
    error(errorMsg,expr.get_SourcePositionStart());
  }


  public Either<Integer, String> calculateNumericValue(ASTExpr expr) {
    //TODO write tests for this
    if (expr.isLeftParentheses()) {
      return calculateNumericValue(expr.getExpr().get());
    }
    else if (expr.getNumericLiteral().isPresent()) {
      if (expr.getNumericLiteral().get() instanceof ASTIntLiteral) {
        ASTIntLiteral literal = (ASTIntLiteral) expr.getNumericLiteral().get();
        return Either.value(literal.getValue());
      }
      else {
        return Either.error(ERROR_CODE+ " " + AstUtils.print(expr.get_SourcePositionStart()) + " : " +
            "No floating point values allowed in the exponent to a UNIT base");
      }
    }
    else if (expr.isUnaryMinus()) {
      Either<Integer, String> term = calculateNumericValue(expr.getTerm().get());
      if (term.isError()) {
        return term;
      }
      return Either.value(-term.getValue());
    }

    return Either.error(ERROR_CODE+ " " + AstUtils.print(expr.get_SourcePositionStart()) + " : " +
        "Cannot calculate value of exponent. Must be a static value!");
  }
}

File: pynestml/nestml/PowVisitor.py
#
# PowVisitor.py
#
# This file is part of NEST.
#
# Copyright (C) 2004 The NEST Initiative
#
# NEST is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 2 of the License, or
# (at your option) any later version.
#
# NEST is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with NEST.  If not, see <http://www.gnu.org/licenses/>.

"""
expression : <assoc=right> left=expression powOp='**' right=expression
"""
from pynestml.nestml.ASTExpression import ASTExpression
from pynestml.nestml.ASTSimpleExpression import ASTSimpleExpression
from pynestml.nestml.PredefinedTypes import PredefinedTypes
from pynestml.nestml.TypeChecker import TypeChecker
from pynestml.nestml.ErrorStrings import ErrorStrings
from pynestml.nestml.NESTMLVisitor import NESTMLVisitor
from pynestml.nestml.Either import Either
from pynestml.utils.Logger import Logger, LOGGING_LEVEL


class PowVisitor(NESTMLVisitor):
    """
    Visits a single power expression and updates its types accordingly.
    """

    def visitExpression(self, _expr=None):
        assert _expr is not None
        baseTypeE = _expr.getLhs().getTypeEither()
        exponentTypeE = _expr.getRhs().getTypeEither()

        if baseTypeE.isError():
            _expr.setTypeEither(baseTypeE)
            return

        if exponentTypeE.isError():
            _expr.setTypeEither(exponentTypeE)
            return

        baseType = baseTypeE.getValue()
        exponentType = exponentTypeE.getValue()

        if TypeChecker.isNumeric(baseType) and TypeChecker.isNumeric(exponentType):
            if TypeChecker.isInteger(baseType) and TypeChecker.isInteger(exponentType):
                _expr.setTypeEither(Either.value(PredefinedTypes.getIntegerType()))
                return
            elif TypeChecker.isUnit(baseType):
                # exponents to units MUST be integer and calculable at time of analysis.
                # Otherwise resulting unit is undefined
                if not TypeChecker.isInteger(exponentType):
                    errorMsg = ErrorStrings.messageUnitBase(self, _expr.getSourcePosition())
                    _expr.setTypeEither(Either.error(errorMsg))
                    Logger.logMessage(errorMsg, LOGGING_LEVEL.ERROR)
                    return
                baseUnit = baseType.getSympyUnit()
                exponentValue = self.calculateNumericValue(
                    _expr.getRhs())  # calculate exponent value if exponent composed of literals
                if exponentValue.isValue():
                    _expr.setTypeEither(
                        Either.value(PredefinedTypes.getTypeIfExists(baseUnit ** exponentValue.getValue())))
                    return
                else:
                    errorMsg = exponentValue.getError()
                    _expr.setTypeEither(Either.error(errorMsg))
                    Logger.logMessage(errorMsg, LOGGING_LEVEL.ERROR)
                    return
            else:
                _expr.setTypeEither(Either.value(PredefinedTypes.getRealType()))
                return
        # Catch-all if no case has matched
        errorMsg = ErrorStrings.messageUnitBase(self, _expr.getSourcePosition())
        _expr.setTypeEither(Either.error(errorMsg))
        Logger.logMessage(errorMsg, LOGGING_LEVEL.ERROR)

    def calculateNumericValue(self, _expr=None):
        # TODO write tests for this
        if isinstance(_expr, ASTExpression) and _expr.isEncapsulated():
            return self.calculateNumericValue(_expr.getExpr())
        elif isinstance(_expr, ASTSimpleExpression) and _expr.getNumericLiteral() is not None:
            if isinstance(_expr.getNumericLiteral(), int):
                literal = _expr.getNumericLiteral()
                return Either.value(literal)
            else:
                errorMessage = ErrorStrings.messageUnitBase(self, _expr.getSourcePosition())
                return Either.error(errorMessage)
        elif _expr.isUnaryOperator() and _expr.getUnaryOperator().isUnaryMinus():
            term = self.calculateNumericValue(_expr.getExpression)
            if term.isError():
                return term
            return Either.value(-term.getValue())
        errorMessage = ErrorStrings.messageNonConstantExponent(self, _expr.getSourcePosition)
        return Either.error(errorMessage)
